src/test/java/org/jenkinsci/plugins/workflow/cps/steps/ingroovy/GroovyStepTest.java

```diff
@@ -0,0 +1,286 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static hudson.model.Result.*;
+import static org.junit.Assert.*;
+
+/**
+ * @author Kohsuke Kawaguchi
+ * @see "doc/step-in-groovy.md"
+ */
+public class GroovyStepTest {
+    @Rule public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    /**
+     * Invokes helloWorldGroovy.groovy in resources/stepsInGroovy and make sure that works.
+     *
+     * Restart Jenkins in the middle to make sure the resurrection works as expected.
+     */
+    @Test
+    public void helloWorld() throws Exception {
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.createProject(WorkflowJob.class, "demo");
+                p.setDefinition(new CpsFlowDefinition(
+                        "helloWorldGroovy('Duke') {\n" +
+                            "echo 'Hello body'\n"+
+                            "semaphore 'restart'\n"+
+                            "echo 'Good bye body'\n"+
+                        "}",true));
+                WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+                SemaphoreStep.waitForStart("restart/1", b);
+                story.j.waitForMessage("Hello body",b);
+                story.j.assertLogContains("Hello Duke",b);
+            }
+        });
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("demo", WorkflowJob.class);
+                WorkflowRun b = p.getLastBuild();
+                SemaphoreStep.success("restart/1", null);
+                story.j.waitForCompletion(b);
+
+                story.j.assertLogContains("Good bye body",b);
+                story.j.assertLogContains("Good bye Duke",b);
+            }
+        });
+    }
+
+    /**
+     * Invokes helloWorldGroovy.groovy in resources/stepsInGroovy and make sure that works.
+     *
+     * Restart Jenkins in the middle to make sure the resurrection works as expected.
+     */
+    @Test
+    public void security() throws Exception {
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.createProject(WorkflowJob.class, "demo");
+
+                // Groovy step should have access to the money in the vault
+                p.setDefinition(new CpsFlowDefinition(
+                        "assert bankTeller()==2000",
+                        true));
+                story.j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+
+                // but no direct access allowed
+                String className = BankTellerSecureStepExecution.class.getName().replace("Secure","");
+                p.setDefinition(new CpsFlowDefinition(
+                        "def o = new "+ className +"()\n"+
```

Comment:
I think it would be better to simply blacklist the constructors of `GroovyStepExecution` subtypes, so you could not get even this far. After all there is no legitimate reason for untrusted code to be calling this. Then there would be no need for script authors to do the `BankTellerSecureStepExecution` trick, which I fear is too easily neglected.